outlines/outline.py

```diff
@@ -0,0 +1,68 @@
+import json
+from dataclasses import dataclass
+
+import jsonschema
+from pydantic import BaseModel
+
+from outlines import generate
+
+
+@dataclass
+class Outline:
+    """
+    Outline is a class that creates a callable object to generate responses
+    based on a given model, a prompt template (a function that returns a `str`) and an expected output type.
+
+    Parameters
+    ----------
+    model : object
+        The model to be used for generating responses.
+    template : function
+        A function that takes arguments and returns a prompt string.
+    output_type : type
+        The expected output type of the generated response.
+
+    Examples
+    --------
+    from pydantic import BaseModel
+    from outlines import models, Outline
+
+    class OutputModel(BaseModel):
+        result: int
+
+    model = models.transformers("gpt2")
+
+    def template(a: int) -> str:
+        return f"What is 2 times {a}?"
+
+    fn = Outline(model, template, OutputModel)
+
+    result = fn(3)
+    print(result)  # Expected output: OutputModel(result=6)
+    """
+
+    def __init__(self, model, template, output_type):
+        if isinstance(output_type, str):
+            try:
+                jsonschema.Draft7Validator.check_schema(json.loads(output_type))
+            except jsonschema.exceptions.SchemaError as e:
+                raise TypeError(f"Invalid JSON Schema: {e.message}")
+        elif not issubclass(output_type, BaseModel):
+            raise TypeError(
+                "output_type must be a Pydantic model or a valid JSON Schema string"
+            )
+
+        self.template = template
+        self.output_type = output_type
+        self.generator = generate.json(model, output_type)
+
+    def __call__(self, *args, **kwargs):
+        prompt = self.template(*args, **kwargs)
+        response = self.generator(prompt)
+        try:
+            if isinstance(self.output_type, str):
+                return json.loads(response)
+            return self.output_type.model_validate_json(response)
```

Comment:
Let's instantiate the model with the JSON data, like we do [here](https://github.com/dottxt-ai/outlines/blob/03d387893b95f151c74ff6b7d7864e15cb484eee/outlines/generate/json.py#L52) to keep a consistent interface.

Reply:
I got this linter warning:

> The method "parse_raw" in class "BaseModel" is deprecated
> The `parse_raw` method is deprecated; if your data is JSON use `model_validate_json`, otherwise load the data then use `model_validate` instead.Pylance

Reply:
You're right, my bad